sql/catalyst/src/main/scala/org/apache/spark/sql/types/UnionType.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.types
+
+import org.apache.spark.annotation.DeveloperApi
+
+@DeveloperApi
+case class UnionType(types: Seq[DataType]) extends DataType {
```

Comment:
Hmm, note that the problem is not with the contents of the table, but with the table metadata.

 (In my original patch I forked Hive's `ComplexType.java` from an old version that didn't include any union columns. But I still needed `UnionType` to parse the table metadata in the first place.)